AWSAuthSDK/Tests/AWSMobileClientTests/AWSMobileClientDeleteUserTests.swift

```diff
@@ -0,0 +1,63 @@
+//
+//  AWSMobileClientSignoutTests.swift
+//  AWSMobileClientTests
+//
+
+import XCTest
+
+import AWSMobileClient
+import AWSAuthCore
+import AWSCognitoIdentityProvider
+
+/// AWSMobileClient tests related to deleteUser operation
+class AWSMobileClientDeleteUserTests: AWSMobileClientTestBase {
+    
+    /// Test successful delete user with callback and on an Authenticated User
+    ///
+    /// - Given: An authenticated session
+    /// - When:
+    ///    - I invoke `deleteUser` with completion callback
+    /// - Then:
+    ///    - My `deleteUser` completion callback is invoked without an error
+    ///    - The user state is `signedOut`
+    ///
+    func testDeleteUser() {
+        let username = "testUser" + UUID().uuidString
+        let signoutExpectation = expectation(description: "Successfully signout")
+        signUpAndVerifyUser(username: username)
+        signIn(username: username)
+        XCTAssertTrue(AWSMobileClient.default().isSignedIn, "Expected to return true for isSignedIn")
+        sleep(1)
+        AWSMobileClient.default().deleteUser { (error) in
+            XCTAssertNil(error, "Received an error from deleteUser: \(String(describing: error))")
+            XCTAssertFalse(AWSMobileClient.default().isSignedIn, "Expected to return false for isSignedIn")
+            signoutExpectation.fulfill()
+        }
+        wait(for: [signoutExpectation], timeout: 2)
+    }
+    
+    /// Test successful delete user with callback on a Unauthenticated user
+    ///
+    /// - Given: An unauthenticated session
+    /// - When:
+    ///    - I invoke `deleteUser` with completion callback
+    /// - Then:
+    ///    - My `deleteUser` completion callback is invoked with a AWSMobileClientError.notSignedIn error.
+    ///
+    func testDeleteUserUnauthenticatedUser() {
+        let username = "testUser" + UUID().uuidString
+        let signoutExpectation = expectation(description: "Successfully signout for unauthenticated User")
+        signUpAndVerifyUser(username: username)
+        XCTAssertTrue(AWSMobileClient.default().isSignedIn == false, "Expected to return false for isSignedIn")
+        sleep(1)
```

Comment:
Is there a way to replace the delay with a completion handler instead of using `sleep`? What you could do is add an optional closure property which is internal and only set for tests which is called once `signUpAndVerifyUser` is done.

Reply:
Agreed, in my opinion this should definitely be handled via a completion callback instead of a sleep.  With that said, the function this is using is shared and handled via a sleep throughout the AWSMobileClient integration tests, so if we change it, we should probably change it throughout, which is beyond the scope of this PR.  We should consider a follow-on PR to change the tests to callbacks throughout AWSMobileClient.

Reply:
👍